### PR TITLE
fix: Hair and facial hair "teleporting" when mobile dies several times

### PR DESCRIPTION
**Pull request by mdodkins**
### Summary

* Added World.NewVirtual for creating virtual serial numbers
* Reserved range 0x7EEEEEEE to 0x7FFFFFFF for virtual serials
* Hair and Facial hair (for mobiles) now use virtual serials instead of FakeSerial() functions
* Consolidated virtual hair to a single `VirtualHairInfo` class.

Corpse hair and facial hair now persists across save/load and hair and facial hair no longer teleport to newest corpse.

### CODE REVIEW

File: Projects/UOContent/Items/Misc/Corpses/CorpseFacialHair.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
+using ModernUO.Serialization;
+
+namespace Server.Items.Misc.Corpses;
+
+[SerializationGenerator(0, false)]
+public partial class CorpseFacialHair
+{
+    [SerializableField(0)]
+    int _itemID;
```
Comment (kamronbatman) [inline Projects/UOContent/Items/Misc/Corpses/CorpseFacialHair.cs:9]:
```suggestion
    private int _itemID;
```

File: Projects/UOContent/Items/Misc/Corpses/CorpseFacialHair.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+    [SerializableField(1)]
+    int _hue;
```
Comment (kamronbatman) [inline Projects/UOContent/Items/Misc/Corpses/CorpseFacialHair.cs:12]:
```suggestion
    private int _hue;
```

File: Projects/UOContent/Items/Misc/Corpses/CorpseHair.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
+using ModernUO.Serialization;
+
+namespace Server.Items.Misc.Corpses;
+
+[SerializationGenerator(0, false)]
+public partial class CorpseHair
+{
+    [SerializableField(0)]
+    int _itemID;
```
Comment (kamronbatman) [inline Projects/UOContent/Items/Misc/Corpses/CorpseHair.cs:9]:
```suggestion
    private int _itemID;
```

File: Projects/UOContent/Items/Misc/Corpses/CorpseHair.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+    [SerializableField(1)]
+    int _hue;
```
Comment (kamronbatman) [inline Projects/UOContent/Items/Misc/Corpses/CorpseHair.cs:12]:
```suggestion
    private int _hue;
```

File: Projects/UOContent/Items/Misc/Corpses/CorpseHair.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+    [DirtyTrackingEntity]
+    public Corpse Owner { get; set; }
+
+    public CorpseHair(Corpse owner, int itemID=0, int hue=0)
```
Comment (kamronbatman) [inline Projects/UOContent/Items/Misc/Corpses/CorpseHair.cs:17]:
```suggestion
    public CorpseHair(Corpse owner, int itemID = 0, int hue = 0)
```

File: Projects/Server/Mobiles/Mobile.cs [hunk 6]
```diff
@@ -273,7 +273,11 @@ public static void Configure()
     private TimerExecutionToken _expireAggrTimerToken;
     private TimerExecutionToken _expireCombatantTimerToken;
     private TimerExecutionToken _expireCriminalTimerToken;
-    private FacialHairInfo m_FacialHair;
+    private VirtualFacialHairInfo m_FacialHair;
+    public VirtualFacialHairInfo FacialHair
+    {
+        get => m_FacialHair ??= new VirtualFacialHairInfo(FacialHairItemID, FacialHairHue);
+    }
```
Comment (kamronbatman) [inline Projects/Server/Mobiles/Mobile.cs:280]:
```suggestion
    public VirtualFacialHairInfo FacialHair => m_FacialHair ??= new VirtualFacialHairInfo(FacialHairItemID, FacialHairHue);
```

File: Projects/Server/Mobiles/Mobile.cs [hunk 7]
```diff
@@ -282,7 +286,11 @@ public static void Configure()
     private BaseGuild m_Guild;
     private string m_GuildTitle;
 
-    private HairInfo m_Hair;
+    private VirtualHairInfo m_Hair;
+    public VirtualHairInfo Hair
+    {
+        get => m_Hair ??= new VirtualHairInfo(HairItemID, HairHue);
+    }
```
Comment (kamronbatman) [inline Projects/Server/Mobiles/Mobile.cs:293]:
```suggestion
    public VirtualHairInfo Hair => m_Hair ??= new VirtualHairInfo(HairItemID, HairHue);
```

File: Projects/UOContent/Items/Misc/Corpses/CorpseFacialHair.cs [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+    [DirtyTrackingEntity]
+    public Corpse Owner { get; set; }
+
+    public CorpseFacialHair(Corpse owner, int itemID=0, int hue=0)
```
Comment (kamronbatman) [inline Projects/UOContent/Items/Misc/Corpses/CorpseFacialHair.cs:17]:
```suggestion
    public CorpseFacialHair(Corpse owner, int itemID = 0, int hue = 0)
```

File: Projects/Server/World/World.cs [hunk 9]
```diff
@@ -53,13 +53,36 @@ public static class World
...
+            // Guarantee unique serials without locking
+            uint value;
+            uint newValue;
+
+            do
+            {
+                value = NextVirtualSerial;
+                newValue = (value >= MaxVirtualSerial ? ResetVirtualSerial : value + 1);
+            }
+            while (Interlocked.CompareExchange(ref NextVirtualSerial, value + 1, value) != value);
```
Comment (kamronbatman) [inline Projects/Server/World/World.cs:81]: If all 17.9 million are taken, this will lock up into an infinite loop?

Comment (mdodkins) [reply Projects/Server/World/World.cs:81]: Oh god, I'll fix this 🙈

**Pull request merged**